File: programming/flatten_a_doubly_linked_list.py
# https://leetcode.com/problems/flatten-a-multilevel-doubly-linked-list
# Runtime: 48 ms, faster than 15.50% of Python3 online submissions (-_-')
# Memory Usage: 14.5 MB, less than 100.00% of Python3 online submissions (o_o)
class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        ptr = head
        while ptr:
            if ptr.child is not None:
                n = ptr.next
                the_next = self.flatten(ptr.child)
                the_next.prev = ptr
                ptr.next = the_next
                ptr.child = None
                p = ptr
                while p.next:
                    p = p.next
                if n:
                    n.prev = p
                    p.next = n
            ptr = ptr.next
        
        return head
        

File: programming/leetcode/flatten_a_doubly_linked_list.cpp
// https://leetcode.com/problems/flatten-a-multilevel-doubly-linked-list
// Runtime: 0 ms, faster than 100.00% of C++ online submissions 
// Memory Usage: 7.4 MB, less than 100.00% of C++ online submissions 
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head) {
        Node *ptr = head, *next = NULL;
        while (ptr) {
            if (ptr->child) {
                next = ptr->next;
                ptr->next = flatten(ptr->child);
                ptr->next->prev = ptr;
                ptr->child = NULL;
                while (ptr->next) {
                    ptr = ptr->next;
                }
                if (next) {
                    ptr->next = next;
                    next->prev = ptr;   
                }
            }
            ptr = ptr->next;
        }
        return head;
    }
};